java/java-psi-api/src/com/intellij/psi/util/OrderedBinaryExpression.java

```diff
@@ -0,0 +1,122 @@
+   * @return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   */
+  public static <O extends PsiExpression> OrderedBinaryExpression<O, PsiExpression> from(final PsiExpression node, final Class<O> firstClass) {
+    PsiBinaryExpression expression = as(node, PsiBinaryExpression.class);
+    if (expression == null || firstClass == null) return null;
+    O leftOperand = as(expression.getLOperand(), firstClass);
+    O rightOperand = as(expression.getROperand(), firstClass);
+    if (leftOperand != null && rightOperand == null) {
+      return new OrderedBinaryExpression<>(leftOperand, expression.getOperationTokenType(), PsiUtil.skipParenthesizedExprDown(expression.getROperand()));
+    }
+    if (leftOperand == null && rightOperand != null) {
+      IElementType mirroredOperator = mirrorOperator(expression);
+      return new OrderedBinaryExpression<>(rightOperand, mirroredOperator, PsiUtil.skipParenthesizedExprDown(expression.getLOperand()));
+    }
+    return null;
+  }
+
+  /**
+   * Return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   *
+   * @param <F>          the required expression type
+   * @param <S>          the required expression type
+   * @param node         the supposed binary expression
+   * @param firstClass   the class representing the required expression type
+   * @param secondClass  the class representing the required expression type
+   * @return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   */
+  public static <F extends PsiExpression, S extends PsiExpression> OrderedBinaryExpression<F, S> from(final PsiExpression node, final Class<F> firstClass, final Class<S> secondClass) {
+    PsiBinaryExpression expression = as(node, PsiBinaryExpression.class);
+    if (expression == null || firstClass == null || secondClass == null) return null;
+    F leftFirst = as(expression.getLOperand(), firstClass);
+    S rightSecond = as(expression.getROperand(), secondClass);
+    if (leftFirst != null && rightSecond != null) {
+      return new OrderedBinaryExpression<>(leftFirst, expression.getOperationTokenType(), rightSecond);
+    }
+    IElementType mirroredOperator = mirrorOperator(expression);
+    if (mirroredOperator != null) {
+      F rightFirst = as(expression.getROperand(), firstClass);
+      S leftSecond = as(expression.getLOperand(), secondClass);
+
+      if (rightFirst != null && leftSecond != null) {
+        return new OrderedBinaryExpression<>(rightFirst, mirroredOperator, leftSecond);
+      }
+    }
+    return null;
+  }
+
+  private static <T extends PsiExpression> T as(PsiExpression input, Class<T> aClass) {
+    return tryCast(PsiUtil.skipParenthesizedExprDown(input), aClass);
+  }
+
+  private static IElementType mirrorOperator(final PsiBinaryExpression expression) {
+    if (Arrays.asList(
+      JavaTokenType.AND,
+      JavaTokenType.ANDAND,
```

Comment:
I would not say that ANDAND and OROR are commutative operations. Let's skip them for now?

Reply:
In this review fix, I haven't put those operators aside because if we do so, it won't work at all for those operators (we will have to return `null`) but the user will use it in full knowledge as it is the point of the method. So what do we do?